src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph_RemoveProjectReference.cs

```diff
@@ -0,0 +1,130 @@
+            builder.Remove(projectId);
+
+            return builder.ToImmutable();
+        }
+
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseTransitiveReferencesMapForRemovedProjectReference(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseTransitiveReferencesMap,
+            ProjectId projectId,
+            ProjectId referencedProjectId)
+        {
+            var builder = existingReverseTransitiveReferencesMap.ToBuilder();
+
+            // Invalidate the transitive reverse references from every project previously referenced by the original
+            // project (transitively), except for cases where the removed project reference could not impact the result.
+            foreach (var (project, references) in existingReverseTransitiveReferencesMap)
+            {
+                if (!references.Contains(referencedProjectId))
+                {
+                    // If projectId references project, it isn't through referencedProjectId so the change doesn't
+                    // impact the dependency graph.
+                    //
+                    // Suppose we start with the following graph, and are removing the project reference A->B:
+                    //
+                    //   A -> B -> C
+                    //   A -> D
+                    //
+                    // This case is not hit for project C. The reverse transitive references for C contains B, which is
+                    // the target project of the removed reference. We can see that project C is impacted by this
+                    // removal, and after the removal, project A will no longer be in the reverse transitive references
+                    // of C.
+                    //
+                    // This case is hit for project D. The reverse transitive references for D does not contain B, which
+                    // means project D cannot be "downstream" of the impact of removing the reference A->B. We can see
+                    // that project A will still be in the reverse transitive references of D.
+                    //
+                    // This optimization does not catch all cases. For example, consider the following graph where we
+                    // are removing the project reference A->B:
+                    //
+                    //   A -> B -> D
+                    //   A -> D
+                    //
+                    // For this example, we do not hit this optimization because D contains project B in the set of
+                    // reverse transitive references. Without more complicated checks, we cannot rule out the
+                    // possibility that A may have been removed from the reverse transitive references of D by the
+                    // removal of the A->B reference.
+                    continue;
+                }
+
+                if (references.Contains(projectId))
```

Comment:
So this is the part I didn't get before -- we know at this point that references.Contains(referencedProjectId). But anything that has referencedProjectId in the reverse transitive graph then also had projectId in the reverse transitive graph, right? So this check can't ever fail?

Assuming that simplification happens than this function doesn't use projectId, and is the same algorithm more or less as ComputeNewTransitiveReferencesMapForRemovedProjectReference, which makes sense. Both algorithms just are finding all things that contain one side of the arrow if you will, and tossing everything out. The forward case ignores the 'what's being referenced' (i.e. the target of the arrow) and the other side the source of the arrow.

My confusion earlier was why these algorithms were different, and now I think my answer is 'they really aren't'....

Reply:
I updated them to have the same high-level form, so it should now be clearer.